File: src/ts/src/static-btree/key.ts
/** Attribute B+tree key encoding, comparison and sentinels -- ports
 *  `static_btree/key.rs` (Rust origin) and `src/cpp/src/key.cpp` (conformant
 *  C++ port). See Format Reference -> "Attribute B+tree" in
 *  `docs/superpowers/plans/2026-07-19-native-cpp-core.md`.
 *
 *  Every key kind is a FIXED-WIDTH, little-endian encoding so that entries in
 *  an on-disk node can be strided over without a length prefix. There is no
 *  sign-flip / total-order bit transform for floats: the bytes are the plain
 *  IEEE-754 bit pattern, and `ordered_float` ordering (NaN greatest, NaN ==
 *  NaN, -0.0 == +0.0) is applied only by `compareKeys`, after decode.
 *
 *  Deliberate divergences from the Rust reader, reproduced here on purpose
 *  (documented in the Format Reference under "Known divergences"; also
 *  followed by the C++ and Python ports):
 *
 *   1. `Byte` columns decode as UNSIGNED `u8`. The writer stores `Byte` as
 *      `u8` and builds its index as `MemoryIndex<u8>`, but Rust's OWN reader
 *      decodes that index as `i8` -- so for stored values > 127 Rust's reader
 *      returns a negative number that was never written. This port matches
 *      the WRITER, not Rust's reader bug.
 *   2. `Json`/`Binary` columns are classified here (`str100`) so the query
 *      layer (Task 14) can REJECT index queries against them, matching
 *      Rust's `UnsupportedColumnType`: they are `FixedStringKey<100>` over a
 *      JSON/binary blob, so index hits are near-meaningless without
 *      post-verification, and rejecting is honest.
 *   3. `f32`/`f64` `keyMax` is `+Infinity`, NOT NaN, even though `ordered_float`
 *      sorts NaN strictly greater than `+Infinity`. Consequence: range
 *      queries lowered against this sentinel (`Ge`, `Ne`, ...) SILENTLY
 *      EXCLUDE NaN-keyed features. This is deliberately lossy, reproduced so
 *      results match Rust -- not an oversight.
 *   4. `datetime` `keyMin` is epoch `{ seconds: 0n, nanos: 0 }`, even though
 *      the wire format is a signed `i64` and negative (pre-1970) timestamps
 *      round-trip fine through `encodeKey`/`decodeKey`. Consequence: `Le`/`Ne`
 *      range queries lowered against this sentinel are BLIND to pre-1970
 *      timestamps. Also deliberately lossy, reproduced for parity with Rust.
 */
import { ColumnType } from '../generated/column-type.js'
import { ErrorCode, FcbError } from '../errors.js'
import {
  readF32, readF64, readI16, readI32, readI64, readU16, readU32, readU64, readU8,
  writeF32, writeF64, writeI16, writeI32, writeI64, writeU16, writeU32, writeU64, writeU8,
} from '../le.js'

export type KeyKind =
  | 'u8' | 'i16' | 'u16' | 'i32' | 'u32' | 'i64' | 'u64'
  | 'f32' | 'f64' | 'bool' | 'datetime' | 'str50' | 'str100'

/** The on-disk `DateTime` key shape: `i64 LE` UNIX seconds, then `u32 LE`
 *  subsecond nanos -- 12 bytes total, never a JS `Date` (which cannot hold
 *  nanosecond precision and cannot represent the sentinel year 9999 used by
 *  `keyMax`). */
export interface DateTimeKey {
  seconds: bigint
  nanos: number
}

const utf8Encoder = new TextEncoder()

const KEY_SIZES: Record<KeyKind, number> = {
  u8: 1,
  i16: 2,
  u16: 2,
  i32: 4,
  u32: 4,
  i64: 8,
  u64: 8,
  f32: 4,
  f64: 8,
  bool: 1,
  datetime: 12,
  str50: 50,
  str100: 100,
}

export function keySize(kind: KeyKind): number {
  return KEY_SIZES[kind]
}

/** Column type -> key kind, exactly as the WRITER emits it
 *  (`writer/attr_index.rs:240`, `:272`, `:288`). Uses the `ColumnType` enum,
 *  never a numeric literal: `src/fbs/header.fbs:9-26` numbers `Byte=0 ...
 *  String=11, Json=12, DateTime=13, Binary=14`, and an off-by-one here
 *  silently classifies the wrong column. `str20`/`i8` are defined upstream
 *  but never produced by the writer, so neither appears in `KeyKind`. */
export function keyKindForColumn(t: ColumnType): KeyKind {
  switch (t) {
    case ColumnType.Bool: return 'bool'
    case ColumnType.Byte: return 'u8' // divergence #1: writer stores u8, not i8
    case ColumnType.UByte: return 'u8'
    case ColumnType.Short: return 'i16'
    case ColumnType.UShort: return 'u16'
    case ColumnType.Int: return 'i32'
    case ColumnType.UInt: return 'u32'
    case ColumnType.Long: return 'i64'
    case ColumnType.ULong: return 'u64'
    case ColumnType.Float: return 'f32'
    case ColumnType.Double: return 'f64'
    case ColumnType.String: return 'str50'
    case ColumnType.DateTime: return 'datetime'
    case ColumnType.Json: return 'str100' // divergence #2: query layer rejects (Task 14)
    case ColumnType.Binary: return 'str100' // divergence #2
    default: {
      const exhaustive: never = t
      throw new FcbError(ErrorCode.UnsupportedColumnType, `unknown column type ${String(exhaustive)}`)
    }
  }
}

/** True only for the fixed-width string kinds. Zero-padding means distinct
 *  values can collide on-disk (`'a'` and `'a\0'` produce identical keys), so
 *  every tree hit for these kinds needs a post-filter against the decoded
 *  attribute value (Task 15) before it is trusted. */
export function needsPostFilter(kind: KeyKind): boolean {
  return kind === 'str50' || kind === 'str100'
}

/** Encodes one key value into a freshly-allocated, fixed-width buffer.
 *
 *  String kinds truncate the UTF-8 encoding of `value` at the BYTE level and
 *  zero-pad the remainder -- this can split a multi-byte UTF-8 sequence in
 *  half, which is why `decodeKey` never turns the bytes back into a string
 *  (see its docstring). */
export function encodeKey(kind: KeyKind, value: unknown): Uint8Array {
  const size = keySize(kind)
  const buf = new Uint8Array(size)
  const dv = new DataView(buf.buffer)
  switch (kind) {
    case 'bool':
      writeU8(dv, 0, (value as boolean) ? 1 : 0)
      break
    case 'u8':
      writeU8(dv, 0, value as number)
      break
    case 'i16':
      writeI16(dv, 0, value as number)
      break
    case 'u16':
      writeU16(dv, 0, value as number)
      break
    case 'i32':
      writeI32(dv, 0, value as number)
      break
    case 'u32':
      writeU32(dv, 0, value as number)
      break
    case 'i64':
      writeI64(dv, 0, value as bigint)
      break
    case 'u64':
      writeU64(dv, 0, value as bigint)
      break
    case 'f32':
      writeF32(dv, 0, value as number)
      break
    case 'f64':
      writeF64(dv, 0, value as number)
      break
    case 'datetime': {
      const dt = value as DateTimeKey
      writeI64(dv, 0, dt.seconds)
      writeU32(dv, 8, dt.nanos)
      break
    }
    case 'str50':
    case 'str100': {
      const encoded = utf8Encoder.encode(value as string)
      buf.set(encoded.subarray(0, Math.min(encoded.length, size)))
      break
    }
    default: {
      const exhaustive: never = kind
      throw new FcbError(ErrorCode.InvalidArgument, `unknown key kind ${String(exhaustive)}`)
    }
  }
  return buf
}

/** Fixed-width string kinds decode to the raw padded Uint8Array, NOT to a
 *  string: a truncated key can end mid-UTF-8, and decoding then re-encoding
 *  would change the bytes the tree is ordered by. Treat the result as
 *  display-only if you ever need text out of it -- `new TextDecoder().decode`
 *  is lossy (replaces the broken tail with U+FFFD) but does not throw. */
export function decodeKey(
  kind: KeyKind, dv: DataView, offset: number,
): number | bigint | boolean | Uint8Array | DateTimeKey {
  switch (kind) {
    case 'bool':
      return readU8(dv, offset) !== 0
    case 'u8':
      return readU8(dv, offset)
    case 'i16':
      return readI16(dv, offset)
    case 'u16':
      return readU16(dv, offset)
    case 'i32':
      return readI32(dv, offset)
    case 'u32':
      return readU32(dv, offset)
    case 'i64':
      return readI64(dv, offset)
    case 'u64':
      return readU64(dv, offset)
    case 'f32':
      return readF32(dv, offset)
    case 'f64':
      return readF64(dv, offset)
    case 'datetime':
      return { seconds: readI64(dv, offset), nanos: readU32(dv, offset + 8) }
    case 'str50':
    case 'str100': {
      const size = keySize(kind)
      return new Uint8Array(dv.buffer, dv.byteOffset + offset, size).slice()
    }
    default: {
      const exhaustive: never = kind
      throw new FcbError(ErrorCode.InvalidArgument, `unknown key kind ${String(exhaustive)}`)
    }
  }
}

/** Normalizes a string-kind operand for comparison: `compareKeys` may be
 *  called with either a raw JS string (a query bound, not yet encoded) or a
 *  Uint8Array (a key already decoded off a tree node), so both must produce
 *  the same fixed-width byte sequence before comparing. */
function toStringKeyBytes(kind: KeyKind, v: unknown): Uint8Array {
  return v instanceof Uint8Array ? v : encodeKey(kind, v)
}

/** Compares two decoded (or not-yet-encoded) key values under this kind's
 *  ordering.
 *
 *  Floats special-case NaN FIRST, then fall back to plain `<`/`>`: neither
 *  `===` nor `Object.is` gives `ordered_float` semantics (NaN sorts greatest
 *  and equals itself; `-0.0` equals `+0.0`). Never "simplify" this to a bare
 *  `a - b` or `a < b` -- that silently reintroduces `NaN !== NaN` and loses
 *  the deliberate NaN-greatest rule the on-disk tree was built with.
 *
 *  String kinds compare the UTF-8-encoded bytes UNSIGNED and lexicographic,
 *  never the JS strings themselves: JS `<` is UTF-16 code-unit order, which
 *  disagrees with byte order for non-BMP text (e.g. `'｡' < '\u{10000}'` is
 *  `false` in JS, but the byte comparison the tree was built with says the
 *  opposite). "Simplifying" this to `a < b` on the raw strings would silently
 *  break every query involving a supplementary-plane character. */
export function compareKeys(kind: KeyKind, a: unknown, b: unknown): number {
  switch (kind) {
    case 'bool': {
      const av = a as boolean
      const bv = b as boolean
      if (av === bv) return 0
      return av ? 1 : -1
    }
    case 'u8':
    case 'i16':
    case 'u16':
    case 'i32':
    case 'u32': {
      const av = a as number
      const bv = b as number
      return av < bv ? -1 : av > bv ? 1 : 0
    }
    case 'i64':
    case 'u64': {
      const av = a as bigint
      const bv = b as bigint
      return av < bv ? -1 : av > bv ? 1 : 0
    }
    case 'f32':
    case 'f64': {
      const av = a as number
      const bv = b as number
      const aNaN = Number.isNaN(av)
      const bNaN = Number.isNaN(bv)
      if (aNaN && bNaN) return 0
      if (aNaN) return 1
      if (bNaN) return -1
      // Plain `<`/`>` already treat -0.0 and +0.0 as equal (neither is
      // strictly less than the other), matching ordered_float here for free.
      return av < bv ? -1 : av > bv ? 1 : 0
    }
    case 'datetime': {
      const av = a as DateTimeKey
      const bv = b as DateTimeKey
      if (av.seconds !== bv.seconds) return av.seconds < bv.seconds ? -1 : 1
      return av.nanos < bv.nanos ? -1 : av.nanos > bv.nanos ? 1 : 0
    }
    case 'str50':
    case 'str100': {
      const ab = toStringKeyBytes(kind, a)
      const bb = toStringKeyBytes(kind, b)
      const len = Math.min(ab.length, bb.length)
      for (let i = 0; i < len; i++) {
        const diff = (ab[i] as number) - (bb[i] as number)
        if (diff !== 0) return diff < 0 ? -1 : 1
      }
      return ab.length - bb.length
    }
    default: {
      const exhaustive: never = kind
      throw new FcbError(ErrorCode.InvalidArgument, `unknown key kind ${String(exhaustive)}`)
    }
  }
}

/** Sentinel minimum for range-query lowering (`find_range(MIN, key)`, etc).
 *  `datetime`'s minimum is EPOCH ZERO, not `-Infinity` seconds, even though
 *  the wire format is a signed i64 that round-trips negative seconds fine
 *  through `encodeKey`/`decodeKey` -- deliberate divergence #4: pre-1970
 *  timestamps are invisible to range queries built from this sentinel. */
export function keyMin(kind: KeyKind): unknown {
  switch (kind) {
    case 'bool': return false
    case 'u8': return 0
    case 'i16': return -32768
    case 'u16': return 0
    case 'i32': return -2147483648
    case 'u32': return 0
    case 'i64': return -(2n ** 63n)
    case 'u64': return 0n
    case 'f32':
    case 'f64': return Number.NEGATIVE_INFINITY
    case 'datetime': return { seconds: 0n, nanos: 0 } as DateTimeKey
    case 'str50':
    case 'str100': return new Uint8Array(keySize(kind)) // all-zero
    default: {
      const exhaustive: never = kind
      throw new FcbError(ErrorCode.InvalidArgument, `unknown key kind ${String(exhaustive)}`)
    }
  }
}

/** Sentinel maximum for range-query lowering. `f32`/`f64`'s maximum is
 *  `+Infinity`, NOT NaN, even though `ordered_float` sorts NaN strictly
 *  above `+Infinity` -- deliberate divergence #3: NaN-keyed features are
 *  invisible to range queries built from this sentinel. */
export function keyMax(kind: KeyKind): unknown {
  switch (kind) {
    case 'bool': return true
    case 'u8': return 255
    case 'i16': return 32767
    case 'u16': return 65535
    case 'i32': return 2147483647
    case 'u32': return 4294967295
    case 'i64': return 2n ** 63n - 1n
    case 'u64': return 2n ** 64n - 1n
    case 'f32':
    case 'f64': return Number.POSITIVE_INFINITY
    case 'datetime':
      // Year 9999, matching `static_btree/key.rs`'s `DateTime<Utc>::max_value()`.
      return { seconds: 253402300799n, nanos: 999999999 } as DateTimeKey
    case 'str50':
    case 'str100': return new Uint8Array(keySize(kind)).fill(0xff)
    default: {
      const exhaustive: never = kind
      throw new FcbError(ErrorCode.InvalidArgument, `unknown key kind ${String(exhaustive)}`)
    }
  }
}

File: src/cpp/src/key.cpp
#include <fcb/generated/header_generated.h>
#include <fcb/key.hpp>

#include <algorithm>
#include <cmath>
#include <cstring>
#include <limits>

namespace fcb {

namespace {

bool is_string_kind(KeyKind k) {
    return k == KeyKind::String20 || k == KeyKind::String50 || k == KeyKind::String100;
}

template <typename T> void put_le(std::vector<std::uint8_t>& out, T value) {
    static_assert(std::is_integral<T>::value, "integral only");
    using U = typename std::make_unsigned<T>::type;
    U u = static_cast<U>(value);
    for (std::size_t i = 0; i < sizeof(T); ++i) {
        out.push_back(static_cast<std::uint8_t>((u >> (8 * i)) & 0xFF));
    }
}

template <typename T> T get_le(bytes_view b, std::size_t at = 0) {
    using U = typename std::make_unsigned<T>::type;
    U u = 0;
    for (std::size_t i = 0; i < sizeof(T); ++i) {
        u |= static_cast<U>(b[at + i]) << (8 * i);
    }
    return static_cast<T>(u);
}

/// ordered_float total order: NaN equals itself and sorts above everything.
/// Do NOT use std::less -- NaN would break the tree's ordering invariants.
int cmp_ordered_double(double a, double b) {
    const bool na = std::isnan(a);
    const bool nb = std::isnan(b);
    if (na && nb)
        return 0;
    if (na)
        return 1;
    if (nb)
        return -1;
    if (a < b)
        return -1;
    if (a > b)
        return 1;
    return 0;  // also covers -0.0 == +0.0
}

void require_size(bytes_view b, std::size_t need, const char* what) {
    if (b.size() < need) {
        throw Error(ErrorCode::InvalidAttributeValue, std::string("short key buffer for ") + what);
    }
}

}  // namespace

std::size_t key_serialized_size(KeyKind kind) {
    switch (kind) {
        case KeyKind::Int8:
        case KeyKind::UInt8:
        case KeyKind::Bool:
            return 1;
        case KeyKind::Int16:
        case KeyKind::UInt16:
            return 2;
        case KeyKind::Int32:
        case KeyKind::UInt32:
        case KeyKind::Float32:
            return 4;
        case KeyKind::Int64:
        case KeyKind::UInt64:
        case KeyKind::Float64:
            return 8;
        case KeyKind::DateTime:
            return 12;  // i64 seconds + u32 nanos
        case KeyKind::String20:
            return 20;
        case KeyKind::String50:
            return 50;
        case KeyKind::String100:
            return 100;
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown key kind");
}

#define FCB_KV_INT(NAME, TYPE, KIND, FIELD)          \
    KeyValue KeyValue::NAME(TYPE v) {                \
        KeyValue k;                                  \
        k.kind_ = KeyKind::KIND;                     \
        k.FIELD = static_cast<decltype(k.FIELD)>(v); \
        return k;                                    \
    }

FCB_KV_INT(from_i8, std::int8_t, Int8, i_)
FCB_KV_INT(from_u8, std::uint8_t, UInt8, u_)
FCB_KV_INT(from_i16, std::int16_t, Int16, i_)
FCB_KV_INT(from_u16, std::uint16_t, UInt16, u_)
FCB_KV_INT(from_i32, std::int32_t, Int32, i_)
FCB_KV_INT(from_u32, std::uint32_t, UInt32, u_)
FCB_KV_INT(from_i64, std::int64_t, Int64, i_)
FCB_KV_INT(from_u64, std::uint64_t, UInt64, u_)
#undef FCB_KV_INT

KeyValue KeyValue::from_f32(float v) {
    KeyValue k;
    k.kind_ = KeyKind::Float32;
    k.f_ = static_cast<double>(v);
    return k;
}

KeyValue KeyValue::from_f64(double v) {
    KeyValue k;
    k.kind_ = KeyKind::Float64;
    k.f_ = v;
    return k;
}

KeyValue KeyValue::from_bool(bool v) {
    KeyValue k;
    k.kind_ = KeyKind::Bool;
    k.u_ = v ? 1 : 0;
    return k;
}

KeyValue KeyValue::from_datetime(std::int64_t seconds, std::uint32_t nanos) {
    KeyValue k;
    k.kind_ = KeyKind::DateTime;
    k.i_ = seconds;
    k.u_ = nanos;
    return k;
}

KeyValue KeyValue::from_string(KeyKind kind, const std::string& v) {
    if (!is_string_kind(kind)) {
        throw Error(ErrorCode::UnsupportedColumnType, "from_string on a non-string kind");
    }
    KeyValue k;
    k.kind_ = kind;
    k.str_ = v;  // kept untruncated, for post-filtering
    return k;
}

std::vector<std::uint8_t> encode_key(const KeyValue& v) {
    std::vector<std::uint8_t> out;
    const std::size_t n = key_serialized_size(v.kind_);
    out.reserve(n);

    switch (v.kind_) {
        case KeyKind::Int8:
            put_le<std::int8_t>(out, static_cast<std::int8_t>(v.i_));
            break;
        case KeyKind::UInt8:
            put_le<std::uint8_t>(out, static_cast<std::uint8_t>(v.u_));
            break;
        case KeyKind::Int16:
            put_le<std::int16_t>(out, static_cast<std::int16_t>(v.i_));
            break;
        case KeyKind::UInt16:
            put_le<std::uint16_t>(out, static_cast<std::uint16_t>(v.u_));
            break;
        case KeyKind::Int32:
            put_le<std::int32_t>(out, static_cast<std::int32_t>(v.i_));
            break;
        case KeyKind::UInt32:
            put_le<std::uint32_t>(out, static_cast<std::uint32_t>(v.u_));
            break;
        case KeyKind::Int64:
            put_le<std::int64_t>(out, v.i_);
            break;
        case KeyKind::UInt64:
            put_le<std::uint64_t>(out, v.u_);
            break;
        case KeyKind::Bool:
            out.push_back(v.u_ != 0 ? 1 : 0);
            break;

        case KeyKind::Float32: {
            // Raw IEEE-754 bits, little-endian. NO order-preserving
            // transform: key.rs:323-345 writes the plain bit pattern, and
            // applying the usual sign-flip trick would disagree with every
            // file the reference has written.
            const float f = static_cast<float>(v.f_);
            std::uint32_t bits;
            std::memcpy(&bits, &f, sizeof(bits));
            put_le<std::uint32_t>(out, bits);
            break;
        }
        case KeyKind::Float64: {
            std::uint64_t bits;
            std::memcpy(&bits, &v.f_, sizeof(bits));
            put_le<std::uint64_t>(out, bits);
            break;
        }
        case KeyKind::DateTime:
            put_le<std::int64_t>(out, v.i_);
            put_le<std::uint32_t>(out, static_cast<std::uint32_t>(v.u_));
            break;

        case KeyKind::String20:
        case KeyKind::String50:
        case KeyKind::String100: {
            // Copy min(len, N) BYTES and zero-pad. Truncation is silent and
            // does not respect UTF-8 boundaries (key.rs:483-489), so two
            // distinct strings sharing an N-byte prefix become identical
            // here -- which is why select_attr must post-filter.
            out.assign(n, 0);
            const std::size_t take = std::min(v.str_.size(), n);
            std::memcpy(out.data(), v.str_.data(), take);
            break;
        }
    }
    return out;
}

KeyValue decode_key(KeyKind kind, bytes_view b) {
    const std::size_t n = key_serialized_size(kind);
    require_size(b, n, "decode_key");

    switch (kind) {
        case KeyKind::Int8:
            return KeyValue::from_i8(get_le<std::int8_t>(b));
        case KeyKind::UInt8:
            return KeyValue::from_u8(get_le<std::uint8_t>(b));
        case KeyKind::Int16:
            return KeyValue::from_i16(get_le<std::int16_t>(b));
        case KeyKind::UInt16:
            return KeyValue::from_u16(get_le<std::uint16_t>(b));
        case KeyKind::Int32:
            return KeyValue::from_i32(get_le<std::int32_t>(b));
        case KeyKind::UInt32:
            return KeyValue::from_u32(get_le<std::uint32_t>(b));
        case KeyKind::Int64:
            return KeyValue::from_i64(get_le<std::int64_t>(b));
        case KeyKind::UInt64:
            return KeyValue::from_u64(get_le<std::uint64_t>(b));
        case KeyKind::Bool:
            return KeyValue::from_bool(b[0] != 0);

        case KeyKind::Float32: {
            const std::uint32_t bits = get_le<std::uint32_t>(b);
            float f;
            std::memcpy(&f, &bits, sizeof(f));
            return KeyValue::from_f32(f);
        }
        case KeyKind::Float64: {
            const std::uint64_t bits = get_le<std::uint64_t>(b);
            double d;
            std::memcpy(&d, &bits, sizeof(d));
            return KeyValue::from_f64(d);
        }
        case KeyKind::DateTime:
            return KeyValue::from_datetime(get_le<std::int64_t>(b), get_le<std::uint32_t>(b, 8));

        case KeyKind::String20:
        case KeyKind::String50:
        case KeyKind::String100: {
            // Stop at the first NUL, as to_string_lossy does (key.rs:511).
            std::size_t len = 0;
            while (len < n && b[len] != 0)
                ++len;
            return KeyValue::from_string(kind,
                                         std::string(reinterpret_cast<const char*>(b.data()), len));
        }
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown key kind");
}

int compare_keys(const KeyValue& a, const KeyValue& b) {
    if (a.kind_ != b.kind_) {
        throw Error(ErrorCode::QueryExecutionError, "comparing keys of different kinds");
    }
    switch (a.kind_) {
        case KeyKind::Int8:
        case KeyKind::Int16:
        case KeyKind::Int32:
        case KeyKind::Int64:
            return a.i_ < b.i_ ? -1 : (a.i_ > b.i_ ? 1 : 0);

        case KeyKind::UInt8:
        case KeyKind::UInt16:
        case KeyKind::UInt32:
        case KeyKind::UInt64:
        case KeyKind::Bool:
            return a.u_ < b.u_ ? -1 : (a.u_ > b.u_ ? 1 : 0);

        case KeyKind::Float32:
        case KeyKind::Float64:
            return cmp_ordered_double(a.f_, b.f_);

        case KeyKind::DateTime:
            if (a.i_ != b.i_)
                return a.i_ < b.i_ ? -1 : 1;
            return a.u_ < b.u_ ? -1 : (a.u_ > b.u_ ? 1 : 0);

        case KeyKind::String20:
        case KeyKind::String50:
        case KeyKind::String100: {
            // Compare the ENCODED (truncated, padded) forms, because that is
            // what the tree stores and orders by.
            const auto ea = encode_key(a);
            const auto eb = encode_key(b);
            const int c = std::memcmp(ea.data(), eb.data(), ea.size());
            return c < 0 ? -1 : (c > 0 ? 1 : 0);
        }
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown key kind");
}

KeyValue key_min(KeyKind kind) {
    switch (kind) {
        case KeyKind::Int8:
            return KeyValue::from_i8(std::numeric_limits<std::int8_t>::min());
        case KeyKind::UInt8:
            return KeyValue::from_u8(0);
        case KeyKind::Int16:
            return KeyValue::from_i16(std::numeric_limits<std::int16_t>::min());
        case KeyKind::UInt16:
            return KeyValue::from_u16(0);
        case KeyKind::Int32:
            return KeyValue::from_i32(std::numeric_limits<std::int32_t>::min());
        case KeyKind::UInt32:
            return KeyValue::from_u32(0);
        case KeyKind::Int64:
            return KeyValue::from_i64(std::numeric_limits<std::int64_t>::min());
        case KeyKind::UInt64:
            return KeyValue::from_u64(0);
        case KeyKind::Float32:
            return KeyValue::from_f32(-std::numeric_limits<float>::infinity());
        case KeyKind::Float64:
            return KeyValue::from_f64(-std::numeric_limits<double>::infinity());
        case KeyKind::Bool:
            return KeyValue::from_bool(false);
        // Epoch 0, matching key.rs:242 -- NOT the true i64 minimum. Pre-1970
        // timestamps are therefore invisible to range queries, in both
        // implementations. Reproduced deliberately.
        case KeyKind::DateTime:
            return KeyValue::from_datetime(0, 0);
        case KeyKind::String20:
        case KeyKind::String50:
        case KeyKind::String100:
            return KeyValue::from_string(kind, std::string());
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown key kind");
}

KeyValue key_max(KeyKind kind) {
    switch (kind) {
        case KeyKind::Int8:
            return KeyValue::from_i8(std::numeric_limits<std::int8_t>::max());
        case KeyKind::UInt8:
            return KeyValue::from_u8(std::numeric_limits<std::uint8_t>::max());
        case KeyKind::Int16:
            return KeyValue::from_i16(std::numeric_limits<std::int16_t>::max());
        case KeyKind::UInt16:
            return KeyValue::from_u16(std::numeric_limits<std::uint16_t>::max());
        case KeyKind::Int32:
            return KeyValue::from_i32(std::numeric_limits<std::int32_t>::max());
        case KeyKind::UInt32:
            return KeyValue::from_u32(std::numeric_limits<std::uint32_t>::max());
        case KeyKind::Int64:
            return KeyValue::from_i64(std::numeric_limits<std::int64_t>::max());
        case KeyKind::UInt64:
            return KeyValue::from_u64(std::numeric_limits<std::uint64_t>::max());
        // +inf, matching key.rs:139. NaN sorts ABOVE +inf in the total
        // order, so NaN-keyed features are excluded from range-lowered
        // operators (Ge, Ne). Reproduced deliberately so results match Rust.
        case KeyKind::Float32:
            return KeyValue::from_f32(std::numeric_limits<float>::infinity());
        case KeyKind::Float64:
            return KeyValue::from_f64(std::numeric_limits<double>::infinity());
        case KeyKind::Bool:
            return KeyValue::from_bool(true);
        case KeyKind::DateTime:
            return KeyValue::from_datetime(253402300799LL, 999999999U);
        case KeyKind::String20:
        case KeyKind::String50:
        case KeyKind::String100:
            return KeyValue::from_string(kind, std::string(key_serialized_size(kind), '\xFF'));
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown key kind");
}

KeyKind key_kind_for_column(std::uint8_t column_type) {
    switch (static_cast<::ColumnType>(column_type)) {
        // Byte -> UInt8, deliberately. The writer stores Byte as u8
        // (writer/attribute.rs) and builds MemoryIndex<u8>
        // (writer/attr_index.rs), so it must be read back unsigned or every
        // stored value above 127 comes back negative. Rust now agrees on
        // both paths -- its index reader (reader/attr_query.rs) and its
        // value reader (reader/deserializer.rs) each decode u8 -- so this is
        // no longer a divergence.
        case ::ColumnType::Byte:
            return KeyKind::UInt8;
        case ::ColumnType::UByte:
            return KeyKind::UInt8;
        case ::ColumnType::Bool:
            return KeyKind::Bool;
        case ::ColumnType::Short:
            return KeyKind::Int16;
        case ::ColumnType::UShort:
            return KeyKind::UInt16;
        case ::ColumnType::Int:
            return KeyKind::Int32;
        case ::ColumnType::UInt:
            return KeyKind::UInt32;
        case ::ColumnType::Long:
            return KeyKind::Int64;
        case ::ColumnType::ULong:
            return KeyKind::UInt64;
        case ::ColumnType::Float:
            return KeyKind::Float32;
        case ::ColumnType::Double:
            return KeyKind::Float64;
        case ::ColumnType::String:
            return KeyKind::String50;
        case ::ColumnType::DateTime:
            return KeyKind::DateTime;
        case ::ColumnType::Json:
            return KeyKind::String100;
        case ::ColumnType::Binary:
            return KeyKind::String100;
    }
    throw Error(ErrorCode::UnsupportedColumnType, "unknown column type");
}

}  // namespace fcb
